### DISCUSSION TITLE
Custom Gravity Direction (per Body)

### DISCUSSION THREAD

**Discussion by PhoenixIllusion** [jrouwe/JoltPhysics.js#166] [Q&A] [answered]
I was looking into how to customize gravity for individual bodies to do things either like gravity-wells, Mario Galaxy walkable tiny planets, or having a go-cart racers drive up a vertical section of special wall.

My current logic is to set the mGravityFactor to zero and then to apply a per-frame force equal to (customGravity * mass) to the body.

Is this the best approach with Jolt's logic?
So far, it seems to at least track `X/Y/Z` exactly the same for a trivial test of two bodies, one under built-in gravity and one emulating  gravity via "Force = Mass * Acceleration", via `AddForce`

I do not know if there are any side-effects, where gravity may be used in other calculations and this 'force' being applied every step does not replicate the same behavior, although if it works fine enough for driving a car, that is enough for me. CharacterVirtual pretty much already uses faked gravity for lots of things, so I am not too concerned about that besides interacting with physics bodies on those planets.

I do see that likely I will need to allow infinite 'PitchRoll' on a vehicle and fix it myself per-frame, due to VehicleConstraint's `mWorldUp` using the physicsSystem's gravity directional vector for stabilizing the vehicles https://github.com/jrouwe/JoltPhysics/blob/7edba88d5336047ebae8c6e26a5aa89dd5ae069e/Jolt/Physics/Vehicle/VehicleConstraint.cpp#L169

I was able to run the 'vehicle_wheeled' demo under that logic (with 180 pitch allowed or the car auto-flips itself) and physicsSystem gravity set to positive Z (upside down) and with positive X (gravity horizontal) and still drive around as long as I didn't flip the car via my own steering. I'd put random spheres and cubes on the track to ensure the real world gravity was still set to awkward directions.

Ex: the following maintains a sync'd `Y` coordinate between two bodies, One Real/One Fake Gravity, even after changing the PhysicSystem direction of gravity mid-way through the fall.

```typescript
import initJolt from './js/jolt-physics.wasm-compat.js';

initJolt().then(function (Jolt) {
  let cb = []; // callback list
  initExample(Jolt, () => cb.forEach(f => f()));

  // Create a basic floor
  let floor = createFloor();
  floor.SetFriction(1.0);

  // control body, uses internal physics system gravity
  const position = new Jolt.RVec3(-1, 15, 0)
  const body = createSphere(position, .5, Jolt.EMotionType_Dynamic, LAYER_MOVING, 0xFF0000)

  // test body, uses force-based simulation of custom gravity
  position.SetX(1);
  const shape = new Jolt.SphereShape(.5, null);
  const creationSettings = new Jolt.BodyCreationSettings(shape, position, Jolt.Quat.prototype.sIdentity(), Jolt.EMotionType_Dynamic, LAYER_MOVING);
  creationSettings.mGravityFactor = 0;
  creationSettings.mOverrideMassProperties = Jolt.EOverrideMassProperties_CalculateInertia;
  creationSettings.mMassPropertiesOverride.mMass = 100;
  const body2 = bodyInterface.CreateBody(creationSettings);
  addToScene(body2, 0xff0000);

  let gravity = physicsSystem.GetGravity().Mul(100); // acceleration times mass
  let logCount = 0;
  let i = 0;
  cb.push(() => {
    if(i++ % 2 == 0 && logCount < 20) {
      logCount++;
      const pointA = body.GetPosition().ToString();
      const pointA_y = body.GetPosition().GetY();
      const pointB = body2.GetPosition().ToString();
      const pointB_y = body2.GetPosition().GetY();
      console.log(pointA, pointB, Math.abs(pointA_y-pointB_y));

      if(logCount == 10) {
        let customGravity = new Jolt.RVec3(-0.5, -2, 0).Normalized().Mul(9.81);
        physicsSystem.SetGravity(customGravity);
        gravity = physicsSystem.GetGravity().Mul(100);
      }
    }
    body2.AddForce(gravity);
  });
});
```

**Comment by jrouwe** [comment, marked as answer]
> Is this the best approach with Jolt's logic?

Yes it is (it's also what we're doing), the accumulated force is applied at the same time as gravity so they're mathematically equivalent:

https://github.com/jrouwe/JoltPhysics/blob/7edba88d5336047ebae8c6e26a5aa89dd5ae069e/Jolt/Physics/Body/MotionProperties.inl#L127

> CharacterVirtual pretty much already uses faked gravity for lots of things

Correct, the update function takes the gravity vector as input, which was done on purpose to allow custom gravity for the character (see CharacterSpaceShipTest C++ demo).

> I do see that likely I will need to allow infinite 'PitchRoll' on a vehicle and fix it myself per-frame, due to VehicleConstraint's mWorldUp using the physicsSystem's gravity directional vector for stabilizing the vehicles

The vehicle world up is indeed only used for 2 things:
* Pitch roll limit constraint
* Keeping the motor cycle upright

It would be very simple to add an API to allow setting a custom world up/gravity vector for the vehicle (that you'd need to call every frame if the gravity varies). With that you'd be able to set a pitch/roll limit again. Let me know if you want this.

**Reply by PhoenixIllusion** [reply]
As a feature, I would classify it as a nice-to-have, but it isn't actively dropping any development in my case.

The fact everything else seems to work except those two things by itself was an impressive feat of Jolts functionality.

Without the feature, I wasn't able to brainstorm my wheeled-vehicle style work around for motorcycles, given one-off writing a CPP level workaround seems I'd need to write a custom new VehicleControllerSetting to instantiate my own new subclass of Motorcycle with my the workaround. My subclass hacking idea was just patching the WorldUp during the PreCollide before calling the super::PreCollide. The simpler self-done route would probably be left with a custom build using a CMakeFile with Patch-Command of the current files, if I got that far on my own.

I believe from everything I've seen, this feature would cover the last few things in Jolt Physics regarding artificial gravity as a whole.

**Reply by jrouwe** [reply]
As said, it was simple: https://github.com/jrouwe/JoltPhysics/pull/1107 (I'll probably merge this tomorrow)

**Reply by jrouwe** [reply]
Ok, it's in!

See: https://www.youtube.com/watch?v=AJPS31S6ZO8

**Reply by PhoenixIllusion** [reply]
Thank you. I'd been brainstorming some ways to demo it in my Babylon plugin. I'd just finished my tachometer and exposing gear shifting and speedometer on my existing demos (looking into configuring Abram Tank engine/transmission using the Allison X1100-5)

Probably using a Path3d to create a path mesh and using the same path normals and tangents to construct and chain together convex hulls extending up/down/left/right from the path segments to assign new gravity. Either sensors or use JS to keep track of vehicle convex hull N and also N-1,N+1 and quick check if the user is still in hull N to reassign override gravity to the new hulls gravity.

That should allow me to do this style of roller coaster race tracks 
https://www.mariowiki.com/Anti-gravity

**Comment by PhoenixIllusion** [comment]
Note to anyone who may attempt this artificial gravity logic:

Add `body.IsActive()` checks during prior to applying artificial gravity or use the `bodyInterface` to `void AddForce([Const, Ref] BodyID inBodyID, [Const, Ref] Vec3 inForce, EActivation inActivationMode);` so the system either wakes it or checks IsActive itself, or you will slowly accumulate gravitational forces until something touches and wakes them up, and cause all the touched bodies to slam themselves against the ground and go flying off into space.

If you think you may be encountering the same issue, as a debug-logic prior to touching an object, you can check:

`body.GetAccumulatedForce().Length()`

I'd wondered why things were seeming to explode after a while on my little planet, though I do need to fix some character "jump" issues when the character's south of the equator:

![image](https://github.com/jrouwe/JoltPhysics.js/assets/756488/ffd96aa7-9163-405b-b118-10ebc3b7fec6)

**Reply by PhoenixIllusion** [reply]
The issue with Jump was with regard to the "Not moving away from ground" logic that is in most of the Character Test logic, `(current_vertical_velocity.GetY() - ground_velocity.GetY()) < 0.1f)` . In a non-"straight down" gravity checks, you'd actually want to use the Dot to extract the velocity of both Character and Ground with respect to your custom gravity system, and compare both of those, but some of the checks are just using the "GetY()".

Once updated to extract gravity-direction velocities, I stopped accumulating infinite velocity towards the center of the planet once my character was on the southern hemisphere and could jump around freely.

https://github.com/jrouwe/JoltPhysics.js/assets/756488/32c944af-faa9-4dc1-ae33-1569344a8b2f

**Reply by jrouwe** [reply]
I think that was a nice demo. I've added a similar one to Jolt's samples: https://github.com/jrouwe/JoltPhysics/commit/83a431bf7e4fa3de523f556f4445933ddfd09489
